### PR TITLE
Queues - Allison Northrop - Scrabble

### PR DESCRIPTION
**Pull request by Allison-Northrop**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Feature 	|   Feedback	|
|---	|---	|
|   How do you feel you and your partner did in sharing responsibilities?	|   When we were working together (when I wasn't sick) I thought we shared the responsibilities and navigating pair programming really well.	|
|   Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |  So that it could be used across all classes associated with Scrabble::Scoring.  |
|   Describe an example of a test you wrote for a _nominal case_. |   I wrote a normal case in Scrabble::Scoring when I checked to see if a method returned an Array	|
|   Describe an example of a test you wrote for an _edge case_. |  An edge case I factored in was when I tested to see if an input wasn't a string.  	|
|   What was your final test coverage percentage?	|  I'm still catching up on the simplecov lecture. I'll go back and look into it once I'm fully caught up 	|
|   Is there a specific piece of code you'd like feedback on?	|  Yes! I couldn't figure out how to def draw_tiles(num) on the Scrabble::TileBag operating appropriately. I think I got half way there, but wasn't able to accurately select random tiles num times. I left more comments in the code. Thank you!  	|

### CODE REVIEW

File: lib/scrabble_scoring.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
+module Scrabble
+
+  class Scoring
+
+    LETTERS = {
```
Comment (droberts-sea) [inline lib/scrabble_scoring.rb:5]: This hash of arrays definitely works, but looking up a letter's score ends up being complicated, involving an `each` loop.

Instead, what if you switched it around and had the keys in this array be letters, and the values be the corresponding score. So it would look like

```ruby
LETTERS = {
  'a': 1,
  'b': 3,
  # ...
}
```

Then looking up a letter would be as simple as `LETTERS[letter]`.

File: lib/scrabble_scoring.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        LETTERS.each do |key, array|
+          if array.include?(letter)
+            total << key
+          end
+        end
+      end
+      if word.length != 7
+        return total.inject(:+)
+      else
+        return total.inject(:+) + 50
```
Comment (droberts-sea) [inline lib/scrabble_scoring.rb:32]: Good use of `inject`. Note that `sum` is equivalent to `inject(:+)`, and a little more clear.

File: lib/scrabble_scoring.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      max_score = word_score_hash.select{|word, score| score == word_score_hash.values.max}
+
+      if max_score != 1
+        #if there are multiple words that tie, then choose the shortest word
+        max_score_sorted = max_score.sort_by {|key, value| key.length}
+        max_score.each do |word, score|
+          if max_score.keys.length == 7
+            return word
+          else
+            return max_score_sorted[0][0]
```
Comment (droberts-sea) [inline lib/scrabble_scoring.rb:51]: You return in both branches of this `if/else`. That means your `each` loop will always hit a return during the first iteration - might as well not even use a loop. As an example of input that fails, try `['zzzzzz', 'aaaaaah']`. I would expect `'aaaaaah'` to win since it has 7 letters, but it does not.

What you probably want here is to move the contents of the `else` clause outside the loop entirely. That way you'll loop through looking for 7 letter words, returning the firs you find, or if you don't find any you'll return the shortest word.

File: specs/scrabble_scoring_spec.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+require 'minitest/reporters'
+require 'minitest/skip_dsl'
+require_relative '../lib/scrabble_scoring'
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe "Scrabble Scoring" do
+
+it "creates a scrabble scoring class" do
+  game_1 = Scrabble::Scoring.new
+  game_1.must_be_kind_of Scrabble::Scoring
```
Comment (droberts-sea) [inline specs/scrabble_scoring_spec.rb:11]: I'm not sure if this test is necessary. Since we're not supposed to instantiate `Scrabble::Scoring`, this tests a behavior that we won't ever take advantage of.

File: lib/scrabble_player.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
+require_relative 'scrabble_scoring'
+
+module Scrabble
+  class Player
+    attr_accessor :name
+    def initialize(name)
+      @name = name
+      @played_words = []
+      @total = []
```
Comment (droberts-sea) [inline lib/scrabble_player.rb:9]: I'm curious why you chose to make `total` an array. Since we don't ever need to look at the list of previous scores, I think it would be a lot simpler to just keep track of an integer.

File: specs/scrabble_player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+  it "returns the highest scoring word" do
+    a = Scrabble::Player.new("allison")
+    a.play("yay")
+    a.play("hi")
+    a.highest_scoring_word.must_be "yay"
+  end
+
+  it "adds 50 points if 7 tiles are used in one word" do
+    Scrabble::Scoring.score("ability").must_equal 62
```
Comment (droberts-sea) [inline specs/scrabble_player_spec.rb:65]: Why are you testing `Scrabble::Scoring` here?

File: specs/scrabble_player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    a.play("Oxyphenbutazone")
+    a.play("Oxyphenbutazone")
+    a.won?.must_be true
+  end
+
+  it "returns false if the game has not been won yet" do
+    a = Scrabble::Player.new("allison")
+    a.play("yay")
+    a.won?.must_be false
+  end
```
Comment (droberts-sea) [inline specs/scrabble_player_spec.rb:95]: I'd like to see some more tests around `won?` here. In particular, what happens at 99 and 100 points? Does it switch over as you would expect?

File: lib/scrabble_tilebag.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+#
+require_relative 'scrabble_scoring'
+
+module Scrabble
+  class TileBag
+    attr_accessor :default_tiles
+
+    def initialize
+      @default_tiles = { A: 9, N: 6, B: 2, O: 8, C: 2, P: 2, D: 4, Q: 1, E: 12, R: 6, F: 2, S: 4, G: 3, T: 6, H: 2,
+      U: 4, I: 9,	V: 2, J: 1,	W: 2, K: 1, X: 1, L: 4,	Y: 2, M: 2, Z: 1 }
```
Comment (droberts-sea) [inline lib/scrabble_tilebag.rb:16]: I like this data representation! Makes it very easy to adjust if, for example, you decide there ought to be 9 "N"s. It might make more sense to arrange them in alphabetical order though.

Comment (droberts-sea) [reply lib/scrabble_tilebag.rb:16]: However, I would like to see you keep the default tiles outside the `initialize` method as a constant, and make a copy of it inside `initialize`. Something like

```ruby
class TileBag
  DEFAULT_TILES = { A: 9, B: 2, ... }
  def initialize
    @tiles = []
    DEFAULT_TILES.each do |letter, frequency|
      @tiles += [letter] * frequency
    end
  end
end
```

Then `@tiles` is a big array of all the tiles in the bag, which should make drawing from it much easier.

File: lib/scrabble_tilebag.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    end
+
+      def draw_tiles(num)
+        tiles_drawn = []
```
Comment (droberts-sea) [inline lib/scrabble_tilebag.rb:20]: Your current data representation makes this method much more difficult than it needs to be! See above for details.

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

|  Feature 	|   Feedback |
|---	|---	|
|  **General**	|   	|
|   Answered comprehension questions 	|   Mostly good. Your second answer doesn't quite hit what we're looking for - the reason to make `Scoring` use class methods is so you don't need an instance of `Scoring` to calculate a score, you can just use it as a method by itself.	|
|   Both Teammates contributed to the codebase	|   N/A	|
|   Regular Commits with meaningful commit messages 	|   I would like to see more frequent commits, with more meaningful commit messages. You should be committing whenever you get a piece of functionality done - at least once for each method you write. As an example of a verbose commit messages, where you write `changed the file name` I would like to see `renamed x.rb to y.rb`, so that it's very clear at a glance exactly what has changed.	|
|  **Scoring class**	|   	|
|  `score` method	|   yes	|
|   Uses appropriate data structure to store the letter score	|   see inline comment	|
|   Appropriately handles edge cases	|   I would like to see more explicit coverage of failure cases. What if someone asks you to score a non-string, or an empty string, or a string full of special characters?	|
|   Tests `score` edge cases (one-letter word, 7 letter word)	|   Positive edge cases (1 letter, 7 letters) are handled, but would like to see more coverage of error cases.	|
|  `highest_scoring_word` method	|   yes	|
|   Appropriately handles edge cases	|   see inline comments	|
|   Tests `highest_scoring_word` edge cases (ties, no words, no array)	| No. There's a lot of required logic around ties that isn't tested at all, and I don't see any coverage of failure cases. |
|  **Player class**	|   	|
|  Uses an array to keep track of words played	|   yes	|
|  Uses existing code to create `highest_scoring_word` and `highest_word_score`	|   yes	|
|  Returns `true` or `false` from `won?` method |  yes |
|  Tests edge cases on the `play(word)` method for words that cannot be scored |    As in `score`, I would like to see more coverage of explicit error cases, rather than just a few tests for success cases. |
|  **TileBag class**	|   	|
|  Uses a data structure to keep the set of default tiles in the bag	|   see inline comment	|
|  `draw_tiles` method uses the parameter to draw a set of random tiles |   not really	|
| `tiles_remaining` method returns a count	|   yes	|
|  Tests `draw_tiles` edge cases for different parameter inputs	|   Again, your testing of nominal cases is fine, but I'd like to see more coverage of error cases. What happens when you try to draw a negative number of tiles? That sort of thing.	|

I would like to start out by acknowledging that you did 2 people's worth of project on your own, which is a ton of work!

For the project itself, this is a good start, but I think there's a way to go still. There are two main areas I think you should focus going forward:

1. Data representation. I want you to be more deliberate in your choice of data structure to solve a given problem. Interesting questions include:
  - Should this be an array or a hash? Why?
  - If it's a hash, what are the keys? What are the values?
  - If it's an array, will having it in a certain order make things easier to work with?
2. Edge and failure cases. How do your methods behave when given bogus data? What about at the boundaries between bogus and good (0 vs 1 characters in a word, for example)?

Feel free to reach out to Jamie, Charles or myself if you want some more guidance in either of these areas!
